sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1982,6 +1982,13 @@ class DatasetSuite extends QueryTest
       assert(timezone == "Asia/Shanghai")
     }
   }
+
+  test("SPARK-34072: Fix empty array failed in functions.lit()") {
+    checkAnswer(
+      spark.range(1).select(lit(Array())),
```

Comment:
Can't you create an array such as: 

```scala
scala> lit(Array.empty[Int])
res0: org.apache.spark.sql.Column = []
```

? Matching `java.lang.Object` to `NullType` is counterintuative.

Reply:
The problem is `Literal.create(Array()).dataType` is also `ArrayType(NullType)`. `lit` just lost the `ClassTag`.

Reply:
BTW `HiveInspectors.javaTypeToDataType` also has this pattern `case c: Class[_] if c == classOf[java.lang.Object] => NullType`.

Reply:
@ulysses-you - does `spark.range(1).select(typedLit(Array()))` get you the result you're looking for?  Per the docs, "The difference between the typedLit function and lit is that this function can handle parameterized scala types e.g.: List, Seq and Map."  `lit` isn't normally used for collections.

Reply:
Yes the right way is using `typedLit`. I just consider the issue is easy to resolve so maybe we can support it directly.